sdks/python/apache_beam/runners/direct/sdf_direct_runner.py

```diff
@@ -98,36 +116,45 @@ def __init__(
         DoFnSignature(self.sdf), context=DoFnContext('unused_context'),
         input_args=args_for_invoker, input_kwargs=kwargs_for_invoker)
 
-  def set_step_context(self, step_context):
+    self._step_context = None
+
+  @property
+  def step_context(self):
+    return self._step_context
+
+  @step_context.setter
+  def step_context(self, step_context):
     assert isinstance(step_context, DirectStepContext)
     self._step_context = step_context
 
   def set_process_element_invoker(self, process_element_invoker):
     assert isinstance(process_element_invoker, SDFProcessElementInvoker)
     self._process_element_invoker = process_element_invoker
 
-  def start_bundle(self):
-    # TODO: support start_bundle() method for SDFs.
-    pass
-
-  def finish_bundle(self):
-    # TODO: support finish_bundle() method for SDFs.
-    pass
-
   def process(self, element, timestamp=beam.DoFn.TimestampParam,
-              window=beam.DoFn.WindowParam, * args, **kwargs):
+              window=beam.DoFn.WindowParam, *args, **kwargs):
     if isinstance(element, KeyedWorkItem):
       # Must be a timer firing.
       key = element.encoded_key
     else:
       key, values = element
       values = list(values)
       assert len(values) == 1
+      # Value here will either be a WindowedValue or an ElementAndRestriction
+      # object.
+      # TODO: handle key collisions here.
+      assert len(values) == 1, 'Internal error. Processing of splittable ' \
+                               'DoFn cannot continue since elements did not ' \
+                               'have unique keys.'
       value = values[0]
+      if len(values) != 1:
+        raise ValueError('')
+      assert isinstance(value, (WindowedValue, ElementAndRestriction))
```

Comment:
Is this checking for being a subclass of both at the same time? Is that how windowed values work in Python - multiple inheritance (rather than wrapping like in Java)?